### ISSUE TITLE
SQLAlchemy 1.4 support?

### ISSUE THREAD

**Issue by ibrag8998** [xzkostyan/clickhouse-sqlalchemy#129]
Any plans on supporting sqlalchemy 1.4+?

**Comment by xzkostyan** [comment]
Yes, 1.4 will be supported. But, I can't provide any ETA.

**Comment by xzkostyan** [comment]
The first step of SQLAlchemy 1.4 support is done: all tests are green. All changes are published in [feature-sa-1.4](https://github.com/xzkostyan/clickhouse-sqlalchemy/tree/feature-sa-1.4) branch.

You can install new version directly from github:

```bash
pip install git+https://github.com/xzkostyan/clickhouse-sqlalchemy@feature-sa-1.4#egg=clickhouse-sqlalchemy==0.2.0
```

This branch can be merged into master and new version will be published on PyPI after some testing.

Looking forward to your feedback.

**Comment by xzkostyan** [comment]
Any feedback? Anyone?

**Comment by phpsxg** [comment]
> ```shell
>  git+https://github.com/xzkostyan/clickhouse-sqlalchemy@feature-sa-1.4#egg=clickhouse-sqlalchemy==0.2.0
> ```

`pip install git+https://github.com/xzkostyan/clickhouse-sqlalchemy@feature-sa-1.4#egg=clickhouse-sqlalchemy==0.2.0

Collecting clickhouse-sqlalchemy==0.2.0
  Cloning https://github.com/xzkostyan/clickhouse-sqlalchemy (to revision feature-sa-1.4) to /tmp/pip-install-dzw1kqhj/clickhouse-sqlalchemy_38fcfc3c4dda49769bf140e2d09df68a
  Running command git clone -q https://github.com/xzkostyan/clickhouse-sqlalchemy /tmp/pip-install-dzw1kqhj/clickhouse-sqlalchemy_38fcfc3c4dda49769bf140e2d09df68a
  fatal: unable to access 'https://github.com/xzkostyan/clickhouse-sqlalchemy/': gnutls_handshake() failed: The TLS connection was non-properly terminated.
WARNING: Discarding git+https://github.com/xzkostyan/clickhouse-sqlalchemy@feature-sa-1.4#egg=clickhouse-sqlalchemy==0.2.0. Command errored out with exit status 128: git clone -q https://github.com/xzkostyan/clickhouse-sqlalchemy /tmp/pip-install-dzw1kqhj/clickhouse-sqlalchemy_38fcfc3c4dda49769bf140e2d09df68a Check the logs for full command output.
ERROR: Could not find a version that satisfies the requirement clickhouse-sqlalchemy (unavailable) (from versions: 0.0.1, 0.0.2, 0.0.3, 0.0.4, 0.0.5, 0.0.6, 0.0.7, 0.0.8, 0.0.9, 0.0.10, 0.1.0, 0.1.1, 0.1.2, 0.1.3, 0.1.4, 0.1.5, 0.1.6, 0.1.7)
ERROR: No matching distribution found for clickhouse-sqlalchemy (unavailable)`

**Comment by xzkostyan** [comment]
Something wrong with:
`gnutls_handshake() failed: The TLS connection was non-properly terminated.`

This works for me:
```bash
$ pip -V
pip 21.1.1 from ~/.pyenv/versions/3.8.10/envs/junk-3.8/lib/python3.8/site-packages/pip (python 3.8)
$ pip install git+https://github.com/xzkostyan/clickhouse-sqlalchemy@feature-sa-1.4#egg=clickhouse-sqlalchemy==0.2.0
Looking in indexes: https://pypi.org/simple
Collecting clickhouse-sqlalchemy==0.2.0
  Cloning https://github.com/xzkostyan/clickhouse-sqlalchemy (to revision feature-sa-1.4) to ./pip-install-ufl7fp7k/clickhouse-sqlalchemy_2eebb035fdb64d75a65ad72a298930de
  Running command git clone -q https://github.com/xzkostyan/clickhouse-sqlalchemy /tmp/pip-install-ufl7fp7k/clickhouse-sqlalchemy_2eebb035fdb64d75a65ad72a298930de
  Running command git checkout -b feature-sa-1.4 --track origin/feature-sa-1.4
  Switched to a new branch 'feature-sa-1.4'
  Branch feature-sa-1.4 set up to track remote branch feature-sa-1.4 from origin.
  WARNING: Requested clickhouse-sqlalchemy==0.2.0 from git+https://github.com/xzkostyan/clickhouse-sqlalchemy@feature-sa-1.4#egg=clickhouse-sqlalchemy==0.2.0, but installing version 0.1.6
Collecting sqlalchemy<1.5,>=1.4
  Using cached SQLAlchemy-1.4.29-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.6 MB)
Collecting requests
  Using cached requests-2.26.0-py2.py3-none-any.whl (62 kB)
Collecting clickhouse-driver>=0.1.2
  Using cached clickhouse_driver-0.2.2-cp38-cp38-manylinux2010_x86_64.whl (718 kB)
Collecting pytz
  Using cached pytz-2021.3-py2.py3-none-any.whl (503 kB)
Collecting tzlocal
  Using cached tzlocal-4.1-py3-none-any.whl (19 kB)
Collecting greenlet!=0.4.17
  Using cached greenlet-1.1.2-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (156 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2021.10.8-py2.py3-none-any.whl (149 kB)
Collecting idna<4,>=2.5
  Using cached idna-3.3-py3-none-any.whl (61 kB)
Collecting urllib3<1.27,>=1.21.1
  Using cached urllib3-1.26.7-py2.py3-none-any.whl (138 kB)
Collecting charset-normalizer~=2.0.0
  Using cached charset_normalizer-2.0.9-py3-none-any.whl (39 kB)
Collecting backports.zoneinfo
  Using cached backports.zoneinfo-0.2.1-cp38-cp38-manylinux1_x86_64.whl (74 kB)
Collecting pytz-deprecation-shim
  Using cached pytz_deprecation_shim-0.1.0.post0-py2.py3-none-any.whl (15 kB)
Collecting tzdata
  Using cached tzdata-2021.5-py2.py3-none-any.whl (339 kB)
Using legacy 'setup.py install' for clickhouse-sqlalchemy, since package 'wheel' is not installed.
Installing collected packages: tzdata, backports.zoneinfo, pytz-deprecation-shim, urllib3, tzlocal, pytz, idna, greenlet, charset-normalizer, certifi, sqlalchemy, requests, clickhouse-driver, clickhouse-sqlalchemy
    Running setup.py install for clickhouse-sqlalchemy ... done
Successfully installed backports.zoneinfo-0.2.1 certifi-2021.10.8 charset-normalizer-2.0.9 clickhouse-driver-0.2.2 clickhouse-sqlalchemy-0.1.6 greenlet-1.1.2 idna-3.3 pytz-2021.3 pytz-deprecation-shim-0.1.0.post0 requests-2.26.0 sqlalchemy-1.4.29 tzdata-2021.5 tzlocal-4.1 urllib3-1.26.7
WARNING: You are using pip version 21.1.1; however, version 21.3.1 is available.
You should consider upgrading via the '~/.pyenv/versions/3.8.10/envs/junk-3.8/bin/python3.8 -m pip install --upgrade pip' command.
```

**Comment by phpsxg** [comment]
When I connect to clickhouse asynchronously, the error is as follows:
**sqlalchemy.exc.InvalidRequestError: The asyncio extension requires an async driver to be used. The loaded 'http' is not async.**

`  uri = "clickhouse://default:@{host}:{port}/{database}"
  engine = create_async_engine(uri, echo=True)`

What is the dirive_name of asynchronous clickhouse，thanks

**Comment by xzkostyan** [comment]
Currently there is no pluggable asynchronous driver for clickhouse.
It seems that async SA will not work with clickhouse until this driver implementation.

**Comment by phpsxg** [comment]
> Currently there is no pluggable asynchronous driver for clickhouse. It seems that async SA will not work with clickhouse until this driver implementation.

Because I found that some SQL queries with slow response time are cumulative when they are concurrent, I found that this problem still exists with thread pools, so I plan to compare them with asynchronous.
Do you have any good suggestions to avoid the accumulation of slow query time?

**Comment by mv-yurchenko** [comment]
@xzkostyan Hello, got the following issue with async issue

`ImportError: cannot import name '_literal_as_label_reference' from 'sqlalchemy.sql.elements`

**Comment by xzkostyan** [comment]
@mv-yurchenko please provide stacktrace.

**Comment by mv-yurchenko** [comment]
> @mv-yurchenko please provide stacktrace.

Issue resolved by dependencies setup of new pip version and orm

**Comment by i8enn** [comment]
> Currently there is no pluggable asynchronous driver for clickhouse. It seems that async SA will not work with clickhouse until this driver implementation.

@xzkostyan This driver may be suitable for you - it is an asynchronous wrapper over the [clickhouse-driver](https://github.com/mymarilyn/clickhouse-driver) used here:
- Native: https://github.com/long2ice/asynch
- HTTP-based: https://github.com/maximdanilchenko/aiochclient

Maybe this will help somehow :)

P.S.: HTTP-based driver works with few http-clients: aiohttp or httpx.

**Comment by cpcloud** [comment]
@xzkostyan I'm interested in helping move this along. How can I help?

**Comment by xzkostyan** [comment]
@cpcloud do you mean async of just plain 1.4 support? Basic 1.4 support is ready in `feature-sa-1.4` branch. Feel free to fork project and play with async support.

There is no feedback except async usage. I'm very disappointed. It seems that we just need release 1.4 support and fix bugs after release.

**Comment by cpcloud** [comment]
@xzkostyan I was referring to 1.4 support. We use this project in https://github.com/ibis-project/ibis and later versions of pandas are now requiring sqlalchemy 1.4. Ibis doesn't use `async`, so we're not blocked on support for that.

**Comment by cpcloud** [comment]
@xzkostyan Is there any reason not to cut a new release with SQLAlchemy 1.4 support?

**Comment by xzkostyan** [comment]
@cpcloud I've just been waiting for positive feedback of this branch. Since our projects stuck at 1.3. I'll try to make new release this weekend.

**Comment by xzkostyan** [comment]
Version 0.2.0 with SQLAlchemy 1.4 support is released. Sorry for the delay.

**Comment by xzkostyan** [comment]
@Bykov25 Hi. No, async is not supported.
